### ISSUE TITLE
Generate linguistic overrides for common *_dot_types

### ISSUE THREAD

**Issue by 3v1n0** [twpayne/chezmoi#2506]
## Is your feature request related to a problem? Please describe.

The files stored in the github repos are not properly highlighted because custom naming causes files not to have proper language associated.

## Describe the solution you'd like

As explained in the [linguistic docs](https://github.com/github/linguist/blob/master/docs/overrides.md), can be defined overrides via `.gitattributes`, that would allow to define custom types.

And defining something like this:

```
# See https://github.com/github/linguist/blob/master/docs/overrides.md
*_dot_bash* linguist-language=bash
*_dot_zsh* linguist-language=zsh
```
Would help, but there are many other cases that should be handled. And probably this can be done programmatically, depending on the dot-files of each user.

**Comment by bradenhilton** [comment]
> ```
> *_dot_bash* linguist-language=bash
> *_dot_zsh* linguist-language=zsh
> ```

Aren't they both just aliases for "Shell"?

https://github.com/github/linguist/blob/737d12e24ad5f739934d29216229b8a22336e9ef/lib/linguist/languages.yml#L6181-L6188

If we were to do this, we would need to base it on the target file name and not the source file name.

We would probably also need to strip the `.tmpl` suffix from templates.

**Comment by twpayne** [comment]
This is an interesting request. chezmoi is a personal dotfile manager. Ensuring that GitHub's syntax highlighting and language statistics are accurate is far from chezmoi's core purpose.

I would first like to see this implemented in a separate repo, along the lines of other [software related to chezmoi](https://www.chezmoi.io/links/related-software/). Once it's mature there, then we can consider moving it into either the [chezmoi organization](https://github.com/chezmoi) or into chezmoi itself.

As a suggestion for how to implement this, this should be possible with a bit of scripting around commands like `chezmoi managed` and `chezmoi source-path`. An initial implementation could look something like:

1. Run `chezmoi managed` to get a list of targets.
2. For each target:
    1. Determine what the appropriate `linguist-language` is, based on the target filename.
    2. Run `chezmoi source-path`  to get the source path.
    3. Emit an appropriate `.gitattributes` line.

Obviously this is not perfect, for example `chezmoi managed` will only list targets for the current machine (potentially ignoring other targets that are only manifested on other machines), and generating a `.gitattributes` file based only on chezmoi needs to be merged with any other `.gitattributes` that the user has set.

> We would probably also need to strip the `.tmpl` suffix from templates.

This would generate correct language statistics but incorrect syntax highlighting. `.tmpl` files are Go templates first and whatever the target language second, so highlighting them as files in the target language will lead to incorrect syntax highlighting. As far as I am aware, there is no way to tell GitHub that "this is a shell script generated from a Go template".

In summary, this request is a niche feature that is complex to implement, so I would like to see a reasonable proof of concept before considering it for chezmoi itself.

**Comment by MunifTanjim** [comment]
As far as I can tell, GitHub understands:

- shebang line: https://github.com/MunifTanjim/dotfiles/blob/main/dot_bashrc
- vim modeline: https://github.com/MunifTanjim/dotfiles/blob/main/dot_zshenv

**Comment by twpayne** [comment]
Thanks @MunifTanjim - that's a much better solution, and I think fixes the problem here. Please re-open if needed.

**Comment by halostatue** [comment]
One thought on this, too. If you wanted to have a vim modeline but *not* have it show up in the generated file, you could use a template comment: `{{- /* vim: set ft=zsh */ -}}` (might be worth mentioning in the docs, but I’m not quite sure where).
